volatility3/framework/plugins/windows/pe_symbols.py

```diff
@@ -0,0 +1,777 @@
+        ]
+
+        for symbol_key, symbol_getter in symbol_keys:
+            # address or name
+            if symbol_key in wanted_symbols:
+                # walk each wanted address or name
+                for wanted_value in wanted_symbols[symbol_key]:
+                    symbol_value = symbol_resolver.__getattribute__(symbol_getter)(
+                        wanted_value
+                    )
+                    if symbol_value:
+                        # yield out symbol name, symbol address
+                        if symbol_key == PESymbols.wanted_names:
+                            yield wanted_value, symbol_value  # type: ignore
+                        else:
+                            yield symbol_value, wanted_value  # type: ignore
+
+                        index = wanted_modules[mod_name][symbol_key].index(wanted_value)  # type: ignore
+
+                        del wanted_modules[mod_name][symbol_key][index]
+
+                # if all names or addresses from a module are found, delete the key
+                if not wanted_modules[mod_name][symbol_key]:
+                    del wanted_modules[mod_name][symbol_key]
+                    break
+
+    @staticmethod
+    def _resolve_symbols_through_methods(
+        context: interfaces.context.ContextInterface,
+        config_path: str,
+        module_instances: List[Tuple[str, int, int]],
+        wanted_modules: PESymbolFinder.cached_value_dict,
+        mod_name: str,
+    ) -> Generator[Tuple[str, int], None, None]:
+        """
+        Attempts to resolve every wanted symbol in `mod_name`
+        Every layer is enumerated for maximum chance of recovery
+        """
+        symbol_resolving_methods = [
+            PESymbols._find_symbols_through_pdb,
+            PESymbols._find_symbols_through_exports,
+        ]
+
+        for method in symbol_resolving_methods:
+            for symbol_resolver in method(
+                context, config_path, module_instances, mod_name
+            ):
+                vollog.debug(f"Have resolver for method {method}")
+                yield from PESymbols._get_symbol_value(
+                    wanted_modules, mod_name, symbol_resolver
+                )
+
+                if not wanted_modules[mod_name]:
+                    break
+
+            if not wanted_modules[mod_name]:
+                break
+
+    @staticmethod
+    def find_symbols(
+        context: interfaces.context.ContextInterface,
+        config_path: str,
+        wanted_modules: PESymbolFinder.cached_value_dict,
+        collected_modules: Dict[str, List[Tuple[str, int, int]]],
+    ) -> found_symbols_type:
+        """
+        Loops through each method of symbol analysis until each wanted symbol is found
+        Returns the resolved symbols as a dictionary that includes the name and runtime address
+        """
+        found_symbols: PESymbols.found_symbols_type = {}
+
+        for mod_name in wanted_modules:
```

Comment:
Why are we handing in these two only to separate them out first, why isn't collected_modules just a filtered list of the module we actually want to check?

The caller could easily do `dict([(mod, collected_modules[mod]) for mod in collected_modules if mod in wanted_modules])`.  Having all of collected_modules doesn't seem necessary?

Reply:
In the case where you are going to look up by address, such as debug registers and several other written plugins blocked on this PR, so you will see more usage soon also, you don't know the module ahead of time. So collected_modules gathers the VAD range for each module in each process. That way we can resolve symbols inside of every process to give us the best chance of finding it.

In the case where you know the module, see unhooked_system_calls, then you just send in the ones you specifically want.

Reply:
But the first thing this does is trim them down, I don't understand why they can't be passed in that way by the caller?  wanted_modules is passed in, collected_modules is passed in, why are we taking in both when we literally only use one to filter down the other, rather than just accepting a pre-filtered list.  Or if not, why are we taking so many lines to filter it when that dict line above will do it in 1?

Again, I think it's my aversion to passing in an input parameter that's used as an output parameter.  Really, really not keen on that design pattern.

Reply:
Granted, it just feels like we're asking people to hand us a lot of inputs, and I'm not sure people will understand why they bother.  Lots of parameters is also a design failure, but there's not too many here so it's fine to keep it if you think it's worthwhile.